### ISSUE TITLE
[FSDP2] Issues with model not running on all ranks - Grads not matching fairscale implementation

### ISSUE THREAD

**Issue by JosselinSomervilleRoberts** [pytorch/pytorch#148334]
### 🐛 Describe the bug

Hi, I am running into some issues because I have a model that does not have to run on all ranks.
Here is a minimal example:
```python
model = CombinedModel()
model = fully_shard(model)
for (x0, x1), y in dataloader:
    if x1 is not None:
        x0 += model.encoder(x1)
    y_pred = model(x0)
    loss = criterion(y, y_pred)
    loss.backward()
    optimizer.step()
```

This is a bit tricky because given like this, the code will hang when trying to gather the shard for a layer in the encoder because some ranks will not run the encoder. Is there a good way to do this?
Right now, to solve this, I do a dummy pass:
```python
if x1 is not None:
    x0 += model.encoder(x1)
else:
    _ = model.encoder(dummy)
```

This solves the forward hanging. However the backward will have the same hanging issue. To solve this, I do this trick but please let me know if there is a better way to do this:
```python
if x1 is not None:
    x0 += model.encoder(x1)
else:
    x0 += 0.0 * model.encoder(dummy)
```

Now the issue is that with this code I get different gradients compared to my fairscale implementation (which does not need all this dummy code). As this may be an important detail, my encoder is fully sharded but I do not shard individual layers of the encoder.

My theory is that since I do not have dummy passes on the encoder with fairscale, in the `all_gather`, the reduce op being average, we will only average ranks that do have gradients. Si if only 2/8 ranks ran the encoder, the divider factor will be 2.
With FSDP2, all ranks will have gradients, most of them being 0 because it was a dummy pass. In that case the sum of the gradients would be the same but the divide factor would be 8.

How can I solve this?
Is there a better way to solve the hang as well ? (One annoying thing is that technically some batches could have no encoder need on all ranks but here we will always do a dummy pass)

Thanks!

### Versions

```
python: Python 3.10.13
torch: 2.4.1
GPUs: H100
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @kwen2501 @c-p-i-o

**Comment by mori360** [comment]
If there is missing information at certain ranks in the dataloader, the code will hang as expected. We recommend addressing this issue by modifying the dataloader to ensure that all ranks contain valid data.

@fegin

**Comment by JosselinSomervilleRoberts** [comment]
The data is valid and we cannot change this. We are just mixing different data sources, some that need extra processing, some not. @mori360

**Comment by mori360** [comment]
You may try the dummy input you described above. But the dummy input would affect the gradient calculation when "some that need extra processing, some not.".
@JosselinSomervilleRoberts

**Comment by wzhang313** [comment]
Hi, this is a rather common situation in multimodal training: certain part of the model is only used with certain type of data. If the data type does not match, the part of model's gradient is None.

Currently, fsdp2 assume all ranks in fsdp shard group has same set of parameters that has gradient, which caused hang.

This user case is supported by FSDP1, so this seems to be a regression?

IMO, the solution is relatively simply, we just need to modify the following line:

https://github.com/pytorch/pytorch/blob/a0893475ba91f2b5c71b31af2be6b716b584ce48/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py#L408

currently, it will only include parameters whose gradient is not None in `fsdp_params_with_grads` and the gradient to `unsharded_grads`. 

we just need to add another branch which add zeros as gradient to it.

```
elif fsdp_param.unsharded_param.grad is None:
      fsdp_params_with_grad.append(fsdp_param)
      unsharded_grads.append(
          fsdp_param.unsharded_param.data.new_zeros(
           fsdp_param.unsharded_param.data.shape
         )
     )
```

**Comment by awgu** [comment]
I think that solution is not correct. FSDP assumes SPMD by design. The fact that it works with FSDP1 is relying on its implementation details with autograd and is actually not a desirable property in other cases like LoRA where some/most parameters are frozen. I think the correct solution is on the user side to pass a dummy input.

**Comment by mori360** [comment]
> If we have dummy gradients that are 0.0, then we can sum them or average them with dp_degree being the divisor, and the result should be correct IIUC?

Yeah, thanks for the details, that solves the gradient scaling issue. I meant that the default gradient scaling would be affected.
